File: folkmq/src/main/java/org/noear/folkmq/client/MqTransaction.java
package org.noear.folkmq.client;

import java.io.IOException;

/**
 * 事务
 *
 * @author noear
 * @since 1.1
 */
public interface MqTransaction {
    /**
     * 事务管理id
     */
    String tmid();

    /**
     * 事务提交
     */
    void commit() throws IOException;

    /**
     * 事务回滚
     */
    void rollback() throws IOException;
}

File: folkmq-lang-python/folkmq/folkmq/client/MqTransaction.py
from abc import abstractmethod, ABC

from socketd.utils.StrUtils import StrUtils


class MqTransaction(ABC):
    @abstractmethod
    def tmid(self) -> str:
        """事务管理id"""
        ...

    @abstractmethod
    def binding(self, message: 'MqMessage'):
        """事务绑定"""
        ...

    @abstractmethod
    async def commit(self):
        """事务提交"""
        ...

    @abstractmethod
    async def rollback(self):
        """事务回滚"""
        ...


class MqTransactionImpl(MqTransaction):
    def __init__(self, client: 'MqClientInternal'):
        self._client = client
        self._keyAry:[str] = []
        self._tmid:str = StrUtils.guid()

    def tmid(self) -> str:
        return self._tmid

    def binding(self, message: 'MqMessage'):
        self._keyAry.append(message.get_key())
        message.internal_sender(self._client.name())

    async def commit(self):
        await self._client.publish2(self._tmid, self._keyAry, False)

    async def rollback(self):
        await self._client.publish2(self._tmid, self._keyAry, True)